File: Day20/RevPolishNotion.py
class Solution(object):
    def evalRPN(self, tokens):
        i = 0
        while len(tokens) > 1:
            if tokens[i] in "+-*/":
                a = int(tokens[i - 2])
                b = int(tokens[i - 1])
                if tokens[i] == "+":
                    c = a + b
                elif tokens[i] == "-":
                    c = a - b
                elif tokens[i] == "*":
                    c = a * b
                else:
                    c = int(float(a) / b)
                tokens[i] = str(c)
                del tokens[i - 1]
                del tokens[i - 2]
                i -= 1
            else:
                i += 1
        return int(tokens[0])

if __name__ == "__main__":
    tokens = input("Enter tokens as space-separated values (e.g. 2 1 + 3 *): ").split()
    print(Solution().evalRPN(tokens))

File: Day20/RevPolishNotion.java
import java.util.*;

public class RevPolishNotion {
    public int evalRPN(String[] tokens) {
        Stack<Integer> stack = new Stack<>();
        for (String token : tokens) {
            if (isOperator(token)) {
                int b = stack.pop();
                int a = stack.pop();
                int result = 0;
                switch (token) {
                    case "+": result = a + b; break;
                    case "-": result = a - b; break;
                    case "*": result = a * b; break;
                    case "/": result = a / b; break;
                }
                stack.push(result);
            } else {
                stack.push(Integer.parseInt(token));
            }
        }
        return stack.pop();
    }

    private boolean isOperator(String s) {
        return s.equals("+") || s.equals("-") || s.equals("*") || s.equals("/");
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        System.out.println("Enter tokens as space-separated values (e.g. 2 1 + 3 *):");
        String[] tokens = sc.nextLine().split(" ");
        System.out.println(new RevPolishNotion().evalRPN(tokens));
    }
}
